core/mavlink_commands.cpp

```diff
@@ -129,6 +129,8 @@ void MAVLinkCommands::receive_command_ack(mavlink_message_t message)
 
     // LogDebug() << "We got an ack: " << command_ack.command;
 
+    _state_mutex.lock();
```

Comment:
Since this is not a `lock_guard` but just a manual lock here you need to do a `unlock` before the two `return` in `if`.

Reply:
Or alternatively you can change it back to a `lock_guard` which avoids this problem. I'm not sure why I changed to this in the first place, it makes it less safe to do code changes.

Reply:
Sorry, I forgot to add two unlocks more which are needed when the lock is moved up in the code.
It seams that you used the lock directly because then you can unlock it as soon as possible, not just at the end of the function. I'll check for this.

Reply:
By using an unique_lock, we can do both: Unlock it earlier when needed but it is surely unlocked when the function is returning in it's destructor.
I changed the order a bit, so that _state_mutex is unlocked after pop_front() to be sure there is no problem there.

Reply:
Cool thx. I think that's a good solution. And I agree that the ordering should now prevent the deadlock. Still, I'm thinking the `_state_mutex` should probably get a better name since it's always "around" the other mutex.